Add usage limits fields for new discount system

boherm opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add usage limits fields for new discount system.<br><br>**Bonus:** add the ability to add a custom label for disabling_switch component.<br/>**Bonus2:** introduce a DirtyTrait to track modified properties in an object.<br>**Bonus3:** add tests for usage limits in FO (and validate the `null` state too for no limit!)
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | See #38663 & #38657
| UI Tests          | https://github.com/boherm/ga.tests.ui.pr/actions/runs/20229798509
| Fixed issue or discussion?     | Fixes #38663 Fixes #38657
| Related PRs       | https://github.com/PrestaShop/ps_apiresources/pull/119
| Sponsor company   | PrestaShop SA

boherm committed:
Add custom label for disabling switch type

boherm committed:
Add Quantity Total & Quantity per customer for discount in usability type

ps-jarvis commented:
<!-- PR_WORDING -->
This pull request seems to contain new translation strings. I have summarized them below to ease up review:

- `Admin.Catalog.Feature`
    - [ ] `Specifiy discount mode`
    - [ ] `Select total usage limits`
    - [ ] `No limit`
    - [ ] `Select usage limits per customer`
- `Admin.Catalog.Help`
    - [ ] `This discount can be used "X" time(s) in total.`
    - [ ] `A customer will only be able to use this discount "X" time(s).`
- `Admin.Catalog.Notification`
    - [ ] `The usage limit per customer must be less than or equal to the total usage limit.`

(Note: this is an automated message, but answering it will reach a real human)

boherm committed:
Make nullable quantity and quantity_per_user for discounts (for no limit!)

boherm committed:
Introduce DirtyTrait to track dirty properties in object

boherm committed:
Add tests for discount quantity & quantity_per_user limit

boherm committed:
Add UnitTests for DirtyTrait

jolelievre commented on classes/CartRule.php:
```diff
@@ -1982,7 +1982,7 @@ public static function autoAddToCart(?Context $context = null, bool $useOrderPri
 		WHERE cr.active = 1
 		AND cr.code = ""
-		AND cr.quantity > 0
+		AND (cr.quantity > 0 OR cr.quantity is null)
```
Did you search in other files (especially legacy ones) if these two fields are used? especially in conditions?

jolelievre commented on src/Adapter/Discount/Update/Filler/DiscountFiller.php:
```diff
@@ -38,59 +38,59 @@ class DiscountFiller
     {
         $updatableProperties = [];
-        if (null !== $command->getValidFrom()) {
+        if ($command->isDirty('validFrom')) {
```
I'm not sure this is required for all use cases 🤔
Technically, it's only meaningful for nullable properties, no?

Not that it impacts in a bad way anyway.

boherm replied:
To me, if we use this new trait, it's better to keep all dirty properties and not only one 🤔

jolelievre commented on src/Core/Domain/Discount/Command/UpdateDiscountCommand.php:
```diff
@@ -94,6 +98,7 @@ public function isActive(): ?bool
     public function setActive(bool $active): self
     {
         $this->active = $active;
+        $this->markDirty('active');
```
Same question as above, do we need to handle the `markDirty` for all fields? Or only the ones that are relevant and have a meaningful null value?

boherm replied:
Same answer as above :D

jolelievre replied:
Alright!

jolelievre commented on src/Core/Domain/Discount/Command/UpdateDiscountCommand.php:
```diff
@@ -133,6 +140,8 @@ public function setValidityDateRange(DateTimeImmutable $from, DateTimeImmutable
         $this->validFrom = $from;
         $this->validTo = $to;
+        $this->markDirty('validFrom');
+        $this->markDirty('validTo');
```
I guess this will be useful for `validTo` that should also be turned into a nullable in another PR

jolelievre commented on src/Core/Trait/DirtyTrait.php:
```diff
@@ -24,45 +24,51 @@
+     * @return void
      */
-    public function __construct(array $tagIds)
+    public function markDirty(string $propertyName): void
```
Out of curiosity, you didn't find a way to automatically mark properties as dirty with `__call` for example? But I guess it's only called when the method doesn't exist right?

boherm replied:
Yep, but it doesn't work. `__call` or `__set` seems to be called only if the function doesn't exist in the class.
So, we need to have here a `markDirty` function to track theses properties.

jolelievre commented on src/PrestaShopBundle/Form/Admin/Sell/Discount/DiscountUsabilityType.php:
```diff
@@ -44,6 +44,59 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                'disabled_value' => function ($data) {
+                    return null === $data;
+                },
+                'disabling_switch_event' => 'quantityTotalSwitchNolimit',
```
This event doesn't seem to be used in JS, maybe it's overkill? (same for the second event)

jolelievre commented on src/PrestaShopBundle/Form/Admin/Sell/Discount/DiscountUsabilityType.php:
```diff
@@ -44,6 +44,59 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                    new Assert\GreaterThanOrEqual(0),
+                    new Assert\LessThanOrEqual([
+                        'propertyPath' => 'parent.all[quantity_total].data',
+                        'message' => $this->trans('The usage limit per customer must be less than or equal to the total usage limit.', 'Admin.Catalog.Notification'),
```
Suggestion for wording, maybe easier to read this way
```suggestion
                        'message' => $this->trans('The usage limit per customer cannot be higher than the total usage limit.', 'Admin.Catalog.Notification'),
```

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/Discount/DiscountFeatureContext.php:
```diff
@@ -92,6 +92,14 @@ public function assertStartDateCannotBeAfterExpiryDate(): void
+    /**
+     * @Then I should get an error that the customer can't use the discount anymore
+     */
+    public function assertDiscountCannotBeInAddedInCart(): void
```
There is already a step for that `I should get an error that the discount is invalid`

boherm committed:
Add ps_apiresources branch for CI

jolelievre reviewed (dismissed):
Thanks

boherm committed:
Hide quantity per customer limit if single_customer selected in discount form

M0rgan01 committed:
handle product customization

M0rgan01 committed:
handle product attribute

boherm committed:
Add 9.1.x branch for github workflows

nicosomb committed:
Merge pull request #40297 from M0rgan01/fix-product-customization-shipment

[multi-carrier] handle product customization

dependabot[bot] committed:
chore(deps-dev): bump @typescript-eslint/parser in /tests/UI

Bumps [@typescript-eslint/parser](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/parser) from 8.48.1 to 8.49.0.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/parser/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v8.49.0/packages/parser)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/parser"
  dependency-version: 8.49.0
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Progi1984 committed:
Merge pull request #40325 from PrestaShop/dependabot/npm_and_yarn/tests/UI/develop/typescript-eslint/parser-8.49.0

chore(deps-dev): bump @typescript-eslint/parser from 8.48.1 to 8.49.0 in /tests/UI

dependabot[bot] committed:
chore(deps-dev): bump @typescript-eslint/eslint-plugin in /tests/UI

Bumps [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/eslint-plugin) from 8.48.1 to 8.49.0.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v8.49.0/packages/eslint-plugin)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/eslint-plugin"
  dependency-version: 8.49.0
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Progi1984 committed:
Merge pull request #40326 from PrestaShop/dependabot/npm_and_yarn/tests/UI/develop/typescript-eslint/eslint-plugin-8.49.0

boherm committed:
Merge pull request #40316 from boherm/add-workflows-for-9.1.x
